### PR TITLE
3838 woodpeckerco

### PR DESCRIPTION
**Pull request by luancazarine**
Triggers
  - Email Sent
  - Email Opened
  - Follow-Up After Autoreply
  - Link Clicked
  - Prospect Autoreplied
  - Prospect Blacklisted
  - Prospect Bounced
  - Prospect Interested
  - Prospect Invalid
  - Prospect Maybe Later
  - Prospect Not Interested
  - Prospect Replied

Actions
  - Create/Update Prospect in Campaign
  - Create/Update Prospect
  - Stop Follow-Ups
  - Find Prospect's Email

<a href="https://gitpod.io/#https://github.com/PipedreamHQ/pipedream/pull/4286"></a>

### CODE REVIEW

**Comment by dylburger** [comment]
Thanks for submitting this PR! When we review PRs, we follow the [Pipedream component guidelines](https://pipedream.com/docs/components/guidelines/). If you're not familiar, here's a quick checklist:

- [ ] [Create components to address specific use cases whenever possible](https://pipedream.com/docs/components/guidelines/#component-scope)
- [ ] [Component `key`s should follow the format `app_name_slug-slugified-component-name`](https://pipedream.com/docs/components/guidelines/#component-key-pattern)
- [ ] Components should follow the [standard directory structure](https://pipedream.com/docs/components/guidelines/#folder-structure)
- [ ] [Prefer Node.js client libraries to REST APIs](https://pipedream.com/docs/components/guidelines/#using-apis-vs-client-libraries)
- [ ] When making API requests, [handle pagination](https://pipedream.com/docs/components/guidelines/#pagination) to ensure all data / events are processed
- [ ] [Use `secret` props](https://pipedream.com/docs/components/guidelines/#capturing-sensitive-data) to capture sensitive data
- [ ] Props and methods should be defined in [app files](https://pipedream.com/docs/components/guidelines/#app-files) whenever possible
- [ ] Document methods with [JS Docs](https://pipedream.com/docs/components/guidelines/#methods)
- [ ] [Use `optional` props](https://pipedream.com/docs/components/guidelines/#optional-vs-required-props) whenever possible, and [set a `default` value](https://pipedream.com/docs/components/guidelines/#default-values) where you can
- [ ] Use [async options](https://pipedream.com/docs/components/guidelines/#async-options) to accept user input wherever possible
- [ ] [Create subscription webhooks sources (vs polling sources) whenever possible](https://pipedream.com/docs/components/guidelines/#webhook-vs-polling-sources)
- [ ] [Always emit a `summary` with events](https://pipedream.com/docs/components/guidelines/#emit-a-summary)
- [ ] [Use a built-in deduping strategy wherever possible](https://pipedream.com/docs/components/guidelines/#deduping)
- [ ] When building webhook sources, [follow these best practices](https://pipedream.com/docs/components/guidelines/#webhook-sources)
- [ ] When building polling sources, [follow these best practices](https://pipedream.com/docs/components/guidelines/#polling-sources)

File: components/woodpecker_co/utils.mjs [hunk 1]
```diff
@@ -0,0 +1,36 @@
+export const orderOptions = [
```
Comment (andrewjschuang) [inline components/woodpecker_co/utils.mjs:1]: Can you put this file inside a `common/` folder so it's not mistaken as a component?

File: components/woodpecker_co/package.json [hunk 2]
```diff
@@ -0,0 +1,22 @@
+{
+  "name": "@pipedream/woodpecker_co",
+  "version": "0.0.2",
```
Comment (andrewjschuang) [inline components/woodpecker_co/package.json:3]:
```suggestion
  "version": "0.0.1",
```

File: components/woodpecker_co/package.json [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+  "description": "Pipedream Woodpecker.co Components",
+  "main": "dist/app/woodpecker_co.app.mjs",
```
Comment (andrewjschuang) [inline components/woodpecker_co/package.json:5]:
```suggestion
  "main": "woodpecker_co.app.mjs",
```

File: components/woodpecker_co/package.json [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+  "keywords": [
+    "pipedream",
+    "woodpecker_co"
+  ],
+  "files": [
+    "dist"
+  ],
```
Comment (andrewjschuang) [inline components/woodpecker_co/package.json:12]:
```suggestion
```

File: components/woodpecker_co/actions/stop-follow-ups/stop-follow-ups.mjs [hunk 5]
```diff
@@ -0,0 +1,49 @@
+import woodpecker from "../../woodpecker_co.app.mjs";
+
+export default {
+  key: "woodpecker_co-stop-follow-ups",
+  name: "Stop Follow-Ups",
+  description: "This action searches a specific prospect. [See the docs here](https://woodpecker.co/help/api-managing-prospects/)",
```
Comment (andrewjschuang) [inline components/woodpecker_co/actions/stop-follow-ups/stop-follow-ups.mjs:6]: Update description

File: components/woodpecker_co/sources/prospect-blacklisted/prospect-blacklisted.mjs [hunk 6]
```diff
@@ -0,0 +1,26 @@
+import common from "../common/base.mjs";
+
+export default {
+  ...common,
+  key: "woodpecker_co-prospect-blacklisted",
+  name: "New Prospect Autoreplied (Instant)",
```
Comment (andrewjschuang) [inline components/woodpecker_co/sources/prospect-blacklisted/prospect-blacklisted.mjs:6]: Update name

File: components/woodpecker_co/woodpecker_co.app.mjs [hunk 7]
```diff
@@ -1,11 +1,164 @@
...
   methods: {
-    // this.$auth contains connected account data
-    authKeys() {
-      console.log(Object.keys(this.$auth));
+    sdk() {
+      return woodpecker(this.$auth.api_key);
+    },
+    async listProspects(params) {
+      params = removeEmpty(params);
+      return await this.sdk().prospects()
```
Comment (andrewjschuang) [inline components/woodpecker_co/woodpecker_co.app.mjs:128]:
```suggestion
      return this.sdk().prospects()
```

File: components/woodpecker_co/woodpecker_co.app.mjs [hunk 8]
```diff
@@ -1,11 +1,164 @@
...
+      const prospects = this.sdk().prospects();
+
+      if (params.id) {
+        return prospects.edit(params, campaignId);
+      } else {
+        return prospects.add(params, campaignId);
+      }
+    },
+    async listCampaigns() {
+      return await this.sdk().campaigns()
```
Comment (andrewjschuang) [inline components/woodpecker_co/woodpecker_co.app.mjs:144]:
```suggestion
      return this.sdk().campaigns()
```

File: components/woodpecker_co/woodpecker_co.app.mjs [hunk 9]
```diff
@@ -1,11 +1,164 @@
...
+        .find();
+    },
+    async unsubscribeProspect() {
+      return await this.sdk().campaigns()
+        .find();
```
Comment (andrewjschuang) [inline components/woodpecker_co/woodpecker_co.app.mjs:149]: Is this right?

File: components/woodpecker_co/woodpecker_co.app.mjs [hunk 10]
```diff
@@ -1,11 +1,164 @@
...
+    },
+    async createHook(
+      url, event,
+    ) {
+      return await this.sdk().webhooks()
```
Comment (andrewjschuang) [inline components/woodpecker_co/woodpecker_co.app.mjs:154]:
```suggestion
      return this.sdk().webhooks()
```

File: components/woodpecker_co/woodpecker_co.app.mjs [hunk 11]
```diff
@@ -1,11 +1,164 @@
...
+        .subscribe(url, event);
+    },
+    async deleteHook(
+      url, event,
+    ) {
+      return await this.sdk().webhooks()
```
Comment (andrewjschuang) [inline components/woodpecker_co/woodpecker_co.app.mjs:160]:
```suggestion
      return this.sdk().webhooks()
```

File: components/woodpecker_co/woodpecker_co.app.mjs [hunk 12]
```diff
@@ -1,11 +1,164 @@
...
+        const prospects = await this.listProspects({
+          campaign: campaignId,
+        });
+        return prospects.map((prospect) => ({
+          value: prospect.id,
+          label: `${prospect.first_name} ${prospect.last_name} (${prospect.email})`,
+        }));
+      },
+    },
+    id: {
```
Comment (andrewjschuang) [inline components/woodpecker_co/woodpecker_co.app.mjs:37]: Is there a way to list the `ids` ?

File: components/woodpecker_co/woodpecker_co.app.mjs [hunk 13]
```diff
@@ -1,11 +1,164 @@
...
+    },
+    async createOrUpdateProspect({
+      params, campaignId = null,
+    }) {
+      params = removeEmpty(params);
+      const prospects = this.sdk().prospects();
+
+      if (params.id) {
+        return prospects.edit(params, campaignId);
+      } else {
```
Comment (andrewjschuang) [inline components/woodpecker_co/woodpecker_co.app.mjs:139]: `else` is not needed here

File: components/woodpecker_co/woodpecker_co.app.mjs [hunk 14]
```diff
@@ -1,11 +1,164 @@
...
+    sdk() {
+      return woodpecker(this.$auth.api_key);
+    },
+    async listProspects(params) {
+      params = removeEmpty(params);
+      return await this.sdk().prospects()
+        .find(params);
+    },
+    async createOrUpdateProspect({
+      params, campaignId = null,
```
Comment (andrewjschuang) [inline components/woodpecker_co/woodpecker_co.app.mjs:132]: Isn't the default `undefined` value for `campaignId` sufficient?

**Review by andrewjschuang — CHANGES_REQUESTED** [review]
Hey @luancazarine, I've added just a few comments.

**Review by andrewjschuang — COMMENTED** [review]
Thanks, LGTM!

**Comment by vunguyenhung** [comment]
Hello @luancazarine, I have tested this PR and there're some test cases failed or needed improvement. 

Please check the test report below for more information
[Woodpecker.co_3838_2138.pdf](https://github.com/PipedreamHQ/pipedream/files/9626836/Woodpecker.co_3838_2138.pdf)

**Review by andrewjschuang — COMMENTED** [review]
Just some optional (but recommended) comments!

**Comment by vunguyenhung** [comment]
Hi everyone, all test cases are passed! Ready for release! 

Test report
[Woodpecker.co_3838_1039.pdf](https://github.com/PipedreamHQ/pipedream/files/9630606/Woodpecker.co_3838_1039.pdf)

File: components/woodpecker_co/woodpecker_co.app.mjs [hunk 15]
```diff
@@ -1,11 +1,160 @@
...
+          campaign: campaignId,
+          $page: page + 1,
+          sort: "id",
+        });
+        return prospects.length
+          ? prospects.map((prospect) => ({
+            value: prospect.id,
+            label: `${prospect.first_name} ${prospect.last_name} (${prospect.email})`,
+          }))
+          : [];
```
[the remark this answers is not in the document]
Comment (luancazarine) [reply components/woodpecker_co/woodpecker_co.app.mjs:40]: image
I tried but it didn't work

**Review by andrewjschuang — APPROVED** [review]

File: components/woodpecker_co/woodpecker_co.app.mjs (hunk 15 above)
[the remark this answers is not in the document]
Comment (andrewjschuang) [reply components/woodpecker_co/woodpecker_co.app.mjs:40]: Weird, it will depend on the returned object from the sdk. But np in leaving it like that.

**Pull request merged**